torch/utils/tensorboard/summary.py

```diff
@@ -303,6 +325,75 @@ def scalar(name, tensor, collections=None, new_style=False, double_precision=Fal
         return Summary(value=[Summary.Value(tag=name, simple_value=scalar)])
 
 
+def tensor_proto(tag, tensor):
+    """Outputs a `Summary` protocol buffer containing the tensor.
+    The generated Summary has a Tensor.proto containing the input Tensor.
+    Args:
+      name: A name for the generated node. Will also serve as the series name in
+        TensorBoard.
+      tensor: Tensor to be converted to protobuf
+    Returns:
+      A tensor protobuf in a `Summary` protobuf.
+    Raises:
+      ValueError: If tensor is too big to be converted to protobuf
+    """
+    if tensor.numel() * tensor.itemsize >= (1 << 31):
+        raise ValueError(
+            "tensor is bigger than protocol buffer's hard limit of 2GB in size"
+        )
+
+    # type maps: torch.Tensor type -> (protobuf type, protobuf val field)
+    type_map = {
+        torch.half: ("DT_HALF", "half_val"),
+        torch.float16: ("DT_HALF", "half_val"),
+        torch.bfloat16: ("DT_BFLOAT", "float_val"),
+        torch.float32: ("DT_FLOAT", "float_val"),
+        torch.float: ("DT_FLOAT", "float_val"),
+        torch.float64: ("DT_DOUBLE", "double_val"),
+        torch.double: ("DT_DOUBLE", "double_val"),
+        torch.int8: ("DT_INT8", "int_val"),
+        torch.uint8: ("DT_UINT8", "int_val"),
+        torch.qint8: ("DT_UINT8", "int_val"),
+        torch.int16: ("DT_INT16", "int_val"),
+        torch.short: ("DT_INT16", "int_val"),
+        torch.int: ("DT_INT32", "int_val"),
+        torch.int32: ("DT_INT32", "int_val"),
+        torch.qint32: ("DT_INT32", "int_val"),
+        torch.complex32: ("DT_COMPLEX32", "scomplex_val"),
+        torch.chalf: ("DT_COMPLEX32", "scomplex_val"),
+        torch.complex64: ("DT_COMPLEX64", "scomplex_val"),
+        torch.cfloat: ("DT_COMPLEX64", "scomplex_val"),
+        torch.bool: ("DT_BOOL", "bool_val"),
+        torch.complex128: ("DT_COMPLEX128", "dcomplex_val"),
+        torch.cdouble: ("DT_COMPLEX128", "dcomplex_val"),
+        torch.uint8: ("DT_UINT8", "uint32_val"),
+        torch.quint8: ("DT_UINT8", "uint32_val"),
+        torch.quint4x2: ("DT_UINT8", "uint32_val"),
+    }
+
+    if tensor.dtype in type_map:
+        proto_val_field = type_map[tensor.dtype][1]
+        tensor_proto_args = {
+            "dtype": type_map[tensor.dtype][0],
+            "dim": [
+                TensorShapeProto.Dim(size=tensor.shape[i]) for i in range(tensor.dim())
+            ],
+            proto_val_field: torch.view_as_real(tensor).tolist()
+            if proto_val_field == "scomplex_val" or proto_val_field == "dcomplex_val"
+            else tensor.reshape(-1).tolist(),
```

Comment:
here a bit more idiomatic and modern might be `.flatten()` instead of reshape(-1)

Reply:
Updated, thanks!

Reply:
Added checks for `numel() == 1` and `numel() == 0`